python/ray/experimental/channel/shared_memory_channel.py

```diff
@@ -697,13 +697,15 @@ def _get_self_actor_id(self) -> str:
         """
         Get the actor ID of the current process. If the current process is the
         aDAG owner (e.g., driver or an actor/task that creates aDAG),
-        use the actor ID of the DAGDriverProxyActor.
+        use the actor ID of the DAGDriverProxyActor. The first element of the
+        reader_and_node_list is always the proxy actor if it's a reader per the
+        creation of reader_and_node_list in CompiledDAGNode.
         """
         actor_id = ray.get_runtime_context().get_actor_id()
-        if self._read_by_adag_driver:
+        if actor_id is None and self._read_by_adag_driver:
```

Comment:
this was actually the only that was wrong with #48100 
this function is called to decide what inner channel to read from, and because it would default to the proxy actor when read by driver, it would keep trying to read from SharedBuffer and never IntraProcess if that existed on the same CompositeChannel